### PR TITLE
pyterm: read space after prompt into prompt

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
Currently, when the prompt is read in `pyterm` the space after it is ignored for the prompt and the output command just adds its own prompt. This leads to the next output always having a leading space, see e.g. this output from `tests/shell` using `RIOT_TERMINAL=pyterm`:

```
make: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT2/tests/shell'
/home/mlenders/Repositories/RIOT-OS/RIOT2/dist/tools/pyterm/pyterm -p "/dev/ttyUSB1" -b "500000"
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2021-02-09 14:47:15,071 # Connect to serial port /dev/ttyUSB1
Welcome to pyterm!
Type '/exit' to exit.
bufsize
2021-02-09 14:47:19,712 # bufsize
2021-02-09 14:47:19,712 # 128
> bufsize
2021-02-09 14:47:21,535 #  bufsize
2021-02-09 14:47:21,536 # 128
>
```

While this isn't necessarily a problem in most cases, it becomes a problem when the prompt is expected and the output of a command is empty. In that case, the space is added to the empty output, making it " ", so the prompt output command is never triggered and the prompt is added to the next command in the log output. To demonstrate I added a command `empty` to `tests/shell` that just does nothing and deactivated the command echoing using `CFLAGS=-DCONFIG_SHELL_NO_ECHO=1`:

```
empty
> empty
empty
bufsize
2021-02-09 14:54:33,753 #  > > 128
>
```

This fixes that problem by also reading the assumed space (we already assume the prompt, so I don't see no harm in that) and if it is not a space to skip the reading of the next char in the next iteration of the reader loop.

### Testing procedure
Using the provided `empty` command in `tests/shell` now yields the expected output:

<details>

```console
$ BOARD=iotlab-m3 RIOT_TERMINAL=pyterm CFLAGS=-DCONFIG_SHELL_NO_ECHO=1 RIOT_CI_BUILD=1 make -j --no-print-directory flash term
Building application "tests_shell" for "iotlab-m3" with MCU "stm32".

   text	   data	    bss	    dec	    hex	filename
  12388	    132	   2368	  14888	   3a28	/home/mlenders/Repositories/RIOT-OS/RIOT/tests/shell/bin/iotlab-m3/tests_shell.elf
/home/mlenders/Repositories/RIOT-OS/RIOT/dist/tools/openocd/openocd.sh flash /home/mlenders/Repositories/RIOT-OS/RIOT/tests/shell/bin/iotlab-m3/tests_shell.elf
### Flashing Target ###
Open On-Chip Debugger 0.10.0+dev-01089-g3bfe49266 (2020-02-26-14:18)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
jtag
Warn : Interface already configured, ignoring
trst_and_srst separate srst_nogate trst_push_pull srst_open_drain connect_assert_srst

Info : clock speed 1000 kHz
Info : JTAG tap: stm32f1x.cpu tap/device found: 0x3ba00477 (mfg: 0x23b (ARM Ltd.), part: 0xba00, ver: 0x3)
Info : JTAG tap: stm32f1x.bs tap/device found: 0x06414041 (mfg: 0x020 (STMicroelectronics), part: 0x6414, ver: 0x0)
Info : stm32f1x.cpu: hardware has 6 breakpoints, 4 watchpoints
Info : stm32f1x.cpu: external reset detected
Info : Listening on port 34845 for gdb connections
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* stm32f1x.cpu       cortex_m   little stm32f1x.cpu       reset

Info : JTAG tap: stm32f1x.cpu tap/device found: 0x3ba00477 (mfg: 0x23b (ARM Ltd.), part: 0xba00, ver: 0x3)
Info : JTAG tap: stm32f1x.bs tap/device found: 0x06414041 (mfg: 0x020 (STMicroelectronics), part: 0x6414, ver: 0x0)
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x080007d0 msp: 0x20000200
Info : device id = 0x10016414
Info : flash size = 512kbytes
auto erase enabled
wrote 14336 bytes from file /home/mlenders/Repositories/RIOT-OS/RIOT/tests/shell/bin/iotlab-m3/tests_shell.elf in 0.641664s (21.818 KiB/s)

verified 12520 bytes in 0.210252s (58.152 KiB/s)

Info : JTAG tap: stm32f1x.cpu tap/device found: 0x3ba00477 (mfg: 0x23b (ARM Ltd.), part: 0xba00, ver: 0x3)
Info : JTAG tap: stm32f1x.bs tap/device found: 0x06414041 (mfg: 0x020 (STMicroelectronics), part: 0x6414, ver: 0x0)
shutdown command invoked
Done flashing
/home/mlenders/Repositories/RIOT-OS/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyUSB1" -b "500000"  
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2021-02-09 15:08:12,883 # Connect to serial port /dev/ttyUSB1
Welcome to pyterm!
Type '/exit' to exit.
empty
> empty
> empty
> bufsize
2021-02-09 15:08:25,954 # 128
```

</details>

Naturally, `tests/shell` should still pass, when run as test

```
make -j -C tests/shell flash test
```

### Issues/PRs references
Found in #15951, where a command with empty output lead to problems in the output parsing.

### CODE REVIEW

**Comment by fjmolinas** [comment]
@miri64 do you think the first commit b663ab0 could be split?

**Comment by fjmolinas** [comment]
> While this isn't necessarily a problem in most cases,

I've definitely found this annoying, had no idea where it was coming from, just thought of it as a pyterm quirk hahaha :)

File: dist/tools/pyterm/pyterm [hunk 1]
```diff
@@ -662,33 +662,43 @@ class SerCmd(cmd.Cmd):
...
-                        pass
-                continue
+            # one of the if branches might have read c already
+            if not c_already_read:
+                try:
+                    c = self._read_char()
+                # try to re-open it with a timeout of 1s otherwise
+                except (serial.SerialException, ValueError):
+                    self._handle_serial_exception()
+                    continue
```
Comment (fjmolinas) [inline dist/tools/pyterm/pyterm:700]: I think this part should be its own commit, it's a different issue no?

File: dist/tools/pyterm/pyterm [hunk 2]
```diff
@@ -700,6 +710,16 @@ class SerCmd(cmd.Cmd):
             elif c == self.serprompt and output == "":
                 sys.stdout.write('%c ' % self.serprompt)
                 sys.stdout.flush()
+                # we now need to read away the space we printed
```
Comment (fjmolinas) [inline dist/tools/pyterm/pyterm:713]: Maybe add a comment above on what is ignore for this whole section?

**Review by fjmolinas — COMMENTED** [review]
Looks good, I would only split up some changes.

File: dist/tools/pyterm/pyterm (hunk 2 above)
Comment (fjmolinas) [reply dist/tools/pyterm/pyterm:713]: So the `> + ' '`

File: dist/tools/pyterm/pyterm (hunk 1 above)
Comment (miri64) [reply dist/tools/pyterm/pyterm:700]: Well it's just some putting of functionalities into functions, that I later reuse for the fix. Can put them into their own commit though, if you insist.

File: dist/tools/pyterm/pyterm (hunk 2 above)
Comment (miri64) [reply dist/tools/pyterm/pyterm:713]: The `>` is not ignored. It is checked in the `c == self.serprompt`.

File: dist/tools/pyterm/pyterm (hunk 1 above)
Comment (miri64) [reply dist/tools/pyterm/pyterm:700]: Done and thus force pushed.

File: dist/tools/pyterm/pyterm (hunk 2 above)
Comment (miri64) [reply dist/tools/pyterm/pyterm]: I updated the comment in the latest force push. Is that what you had in mind?

Comment (miri64) [reply dist/tools/pyterm/pyterm]: Better now?

Comment (miri64) [reply dist/tools/pyterm/pyterm]: It's a bit rambling now, but at least it should be easy to understand.

**Review by fjmolinas — APPROVED** [review]
ACK, please squash

**Comment by fjmolinas** [comment]
I set up tests to be ran as well :)

**Comment by miri64** [comment]
Squashed.
